pkg/controller/scansettingbinding/scansettingbinding_controller.go

```diff
@@ -0,0 +1,496 @@
+	default:
+		scan.ComplianceScanSpec.ScanType = compliancev1alpha1.ScanTypePlatform
+	}
+
+	return nil
+}
+
+func resolveProfileReference(r *ReconcileScanSettingBinding, instance *compliancev1alpha1.ScanSettingBinding, ss *compliancev1alpha1.NamedObjectReference, logger logr.Logger) (*profileReference, error) {
+	var profReference profileReference
+
+	profReference.name = ss.Name
+
+	// Retrieve the unstructured object, fill in the Kind and the apiVersion so that the client knows what to look for
+	o := unstructured.Unstructured{}
+	logger.Info("Resolving ProfileReference", "kind", ss.Kind, "api", ss.APIGroup)
+	// FIXME: apiGroup or apiVersion???
+	o.SetAPIVersion(ss.APIGroup)
+	o.SetKind(ss.Kind)
+
+	key := types.NamespacedName{Namespace: instance.Namespace, Name: ss.Name}
+	err := r.client.Get(context.TODO(), key, &o)
+	if errors.IsNotFound(err) {
+		return nil, common.NewRetriableCtrlErrorWithCustomHandler(func() (reconcile.Result, error) {
+			// This might be a temporary issue in the order the objects are being created
+			if r.recorder != nil {
+				r.recorder.Eventf(
+					instance, corev1.EventTypeWarning, "ScanSettingReferenceError",
+					"ScanSettingBinding reference %s not found", key,
+				)
+			}
+
+			return reconcile.Result{RequeueAfter: requeueAfterDefault, Requeue: true}, nil
+		}, "Tailoring ConfigMap not found")
+	} else if err != nil {
+		logger.Error(err, "error looking up NamedObjectReference", "Setting.Name", ss.Name)
+		return nil, err
+	}
+
+	if o.GetKind() == "Profile" {
+		profReference.profile = &o
+		profReference.tailoredProfile = nil
+	} else if o.GetKind() == "TailoredProfile" {
+		logger.Info("Retrieved a TailoredProfile, must also retrieve a Profile it points to")
+		profReference.tailoredProfile = &o
+
+		if len(o.GetOwnerReferences()) != 1 {
+			return nil, fmt.Errorf("expected that tailored profile has 1 owner, got %d", len(o.GetOwnerReferences()))
+		}
+		owner := o.GetOwnerReferences()[0]
+
+		profileObj := unstructured.Unstructured{}
+		profileObj.SetAPIVersion(owner.APIVersion)
+		profileObj.SetKind(owner.Kind)
+
+		err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: instance.Namespace, Name: owner.Name}, &profileObj)
+		if err != nil {
+			logger.Error(err, "Cannot find Profile", "Profile.Name", owner.Name)
+			return nil, err
+		}
+		logger.Info("object details", "kind", profileObj.GetKind(), "apiVersion", profileObj.GetAPIVersion())
+		profReference.profile = &profileObj
+	} else {
+		return nil, fmt.Errorf("unsupported Kind %s, use one of Profile, TailoredProfile", o.GetKind())
+	}
+
+	// finally retrieve the bundle that the profile points to
+	if profReference.profile == nil {
+		return nil, fmt.Errorf("couldn't resolve profile")
+	}
+	if len(profReference.profile.GetOwnerReferences()) != 1 {
+		return nil, fmt.Errorf("expected that profile has 1 owner, got %d", len(profReference.profile.GetOwnerReferences()))
+	}
+
+	bundleReference := profReference.profile.GetOwnerReferences()[0]
```

Comment:
shouldn't the kind already be properly set in the owner references? I think here you could just verify that it's the appropriate kind, and use a ProfileBundle object straight, as opposed to using an unstructured object.

On the other hand, you probably want to iterate the owner references here, just in case there is more than one.

Reply:
This is also related to the next question about using Unstructured objects. On one hand we use quite generic references in the CRDs to allow for different versions and even kinds to be used, on the other hand we need to "cast" them to typed objects at some point. So the idea was to have a function that just grabs the links regardless of what they are and then another separate one that uses the typed structures with the right version to construct the scan out of the references. If later we support more versions or different kids, we just branch the code that constructs the scan.

About the multiple owner references, what would the processing of other owners do except fail if there are multiple?

Reply:
I think other owners can be ignored. We really just care about a ProfileBundle being the owner here.

Reply:
So, would a check for Kind being a ProfileBundle be OK? Was the explanation why I used the unstructured types sufficient or would you do the type-casting differently?

Reply:
Just checking the kind is fine. And the explanation was sufficient.

Reply:
Done, also for the other ownership (TP->P)